lib/realtime/broadcast_changes/handler.ex

```diff
@@ -0,0 +1,349 @@
+  end
+
+  @impl true
+  def handle_disconnect(state) do
+    Logger.error("Disconnected from the server: #{inspect(state, pretty: true)}")
+
+    {:noreply, %{state | step: :disconnected}}
+  end
+
+  @impl true
+  def handle_data(data, state) when is_keep_alive(data) do
+    %KeepAlive{reply: reply, wal_end: wal_end} = parse(data)
+    wal_end = wal_end + 1
+
+    message =
+      case reply do
+        :now -> standby_status(wal_end, wal_end, wal_end, reply)
+        :later -> hold()
+      end
+
+    {:noreply, message, state}
+  end
+
+  def handle_data(data, state) when is_write(data) do
+    %Write{message: message} = parse(data)
+    message |> decode_message() |> then(&send(self(), &1))
+    {:noreply, [], state}
+  end
+
+  @impl true
+  def handle_info(%Decoder.Messages.Relation{} = msg, state) do
+    %Decoder.Messages.Relation{id: id, namespace: namespace, name: name, columns: columns} = msg
+    %{relations: relations} = state
+    relation = %{name: name, columns: columns, namespace: namespace}
+    relations = Map.put(relations, id, relation)
+    {:noreply, %{state | relations: relations}}
+  end
+
+  def handle_info(%Decoder.Messages.Insert{} = msg, state) do
+    %Decoder.Messages.Insert{relation_id: relation_id, tuple_data: tuple_data} = msg
+    %{buffer: buffer, relations: relations} = state
+
+    case Map.get(relations, relation_id) do
+      %{columns: columns} ->
+        to_broadcast =
+          tuple_data
+          |> Tuple.to_list()
+          |> Enum.zip(columns)
+          |> Map.new(fn
+            {nil, %{name: name}} -> {name, nil}
+            {value, %{name: name, type: "jsonb"}} -> {name, Jason.decode!(value)}
+            {value, %{name: name, type: "bool"}} -> {name, value == "t"}
+            {value, %{name: name}} -> {name, value}
+          end)
+
+        payload = Map.get(to_broadcast, "payload")
+
+        case payload do
+          nil ->
+            {:noreply, state}
+
+          payload ->
+            id = Map.fetch!(to_broadcast, "id")
+
+            to_broadcast =
+              %{
+                topic: Map.fetch!(to_broadcast, "topic"),
+                event: Map.fetch!(to_broadcast, "event"),
+                private: Map.fetch!(to_broadcast, "private"),
+                payload: Map.put(payload, "id", id)
+              }
+
+            buffer = [to_broadcast | buffer]
+            {:noreply, %{state | buffer: buffer}}
+        end
+
+      _ ->
+        log_error("UnknownBroadcastChangesRelation", "Relation ID not found: #{relation_id}")
+        {:noreply, state}
+    end
+  end
+
+  def handle_info(%Decoder.Messages.Commit{}, %{buffer: []} = state) do
+    {:noreply, state}
+  end
+
+  def handle_info(%Decoder.Messages.Commit{}, state) do
+    %{buffer: buffer, tenant_id: tenant_id} = state
+    buffer = Enum.reverse(buffer)
+    tenant = Cache.get_tenant_by_external_id(tenant_id)
+
+    case BatchBroadcast.broadcast(nil, tenant, %{messages: buffer}, true) do
```

Comment:
no strong opinion honestly. it would be less logic so I would be ok with that.

since this is in our side to control, all inserts into realtime.messages should be quite "sane" and probably won't hit any issues